Creates the dashboard UserProficiency table.

ashercodeorg opened the pull request:
This table will be used to track concept-difficulty proficiency on a per user basis.

For a discussion of alternate table designs, see [here](https://docs.google.com/document/d/1TNqTL-W36bM1H7l5gbk6Sb9P_mzFayCP5DyES3PSt2w).

ashercodeorg commented:
CC @laurelfan

Hamms commented on dashboard/app/models/user_proficiency.rb:
```diff
@@ -0,0 +1,66 @@
+
+    num_levels = 0
+    (difficulty_number..MAXIMUM_CONCEPT_DIFFICULTY).each do |d|
+      field_name = concept + "_" + "d" + d.to_s + "_" + "count"
```
I think `"#{concept}_d#{d.to_s}_count"` is preferred

laurelfan replied:
And conveniently you don't need a to_s because string interpolation does
that for you
On Apr 7, 2016 3:47 PM, "Elijah Hamovitz" notifications@github.com wrote:

> In dashboard/app/models/user_proficiency.rb
> https://github.com/code-dot-org/code-dot-org/pull/7748#discussion_r58957967
> :
> 
> > -    functions_with_params
> > -    conditionals
> > -  )
> > -  # The maximum difficulty ranking for a concept.
> > -  # NOTE: This number is tied to DB columns. DO NOT EDIT without a DB migration.
> > -  MAXIMUM_CONCEPT_DIFFICULTY = 5
> >   +
> > -  # Returns the number of levels the user has shown proficiency in for the
> > -  # indiciated concept and difficulty or higher.
> > -  def get_level_count(concept, difficulty_number)
> > -    return 0 if !CONCEPTS.include? concept
> > -    return 0 if !(1..MAXIMUM_CONCEPT_DIFFICULTY).cover? difficulty_number
> >   +
> > -    num_levels = 0
> > -    (difficulty_number..MAXIMUM_CONCEPT_DIFFICULTY).each do |d|
> > -      field_name = concept + "_" + "d" + d.to_s + "_" + "count"
> 
> I think "#{concept}_d#{d.to_s}_count" is preferred
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/7748/files/247d1b0931dca6314b94f51ed5c0f2118ba17158#r58957967

ashercodeorg replied:
Done.

Hamms commented on dashboard/app/models/user_proficiency.rb:
```diff
@@ -0,0 +1,66 @@
+  def basic_proficiency?
+    concept_proficiency_count = 0
+    # Meta-concepts with one sub-concept.
+    %w(sequencing events variables conditionals).each do |concept|
```
In the event that new concepts do get added, we'll need to add them to this list. I don't think that's enough of a worry to justify refactoring, but it'd probably be worth expanding on the `CONCEPTS` comment to that effect

ashercodeorg replied:
That isn't necessarily true. The set of concepts is only partially tied to our computation of basic proficiency. For example, debugging is a concept that doesn't factor into basic proficiency.

Hamms commented on dashboard/db/migrate/20160407160145_create_user_proficiencies.rb:
```diff
@@ -0,0 +1,63 @@
+      t.timestamps null: false
+      t.datetime :last_progress_at
+
+      t.integer :sequencing_d1_count
```
do we wanna add an explicit `default: 0`?

ashercodeorg replied:
Done.

Hamms commented:
couple of minor nits, otherwise LGTM!

ashercodeorg commented on dashboard/app/models/user_proficiency.rb:
```diff
@@ -0,0 +1,67 @@
+
+    num_levels = 0
+    (difficulty_number..MAXIMUM_CONCEPT_DIFFICULTY).each do |d|
+      field_name = "#{concept}_d#{d}_count"
```
This is another error in hound? If so, I'll disable the check locally.

Hamms replied:
Sure looks that way

ashercodeorg commented:
PTA(thorough)L. Note that I added a reference to the users table with a foreign key, changed the migration timestamp (so that I could merge with staging), added default 0 values for counts, made the user_id field not null, etc.

Hamms commented:
LGTM!
